File: 2958.length-of-longest-subarray-with-at-most-k-frequency.py
#
# @lc app=leetcode id=2958 lang=python3
#
# [2958] Length of Longest Subarray With at Most K Frequency
#

# @lc code=start
from collections import Counter
class Solution:
    def maxSubarrayLengsth(self, nums: list[int], k: int) -> int:
        hashmap = Counter()
        left, ans = 0, 0
        for i in range(len(nums)):
            hashmap[nums[i]] += 1
            while hashmap[nums[i]] > k:
                hashmap[nums[left]] -= 1
                left += 1
            ans = max(ans, i - left  + 1)
        return ans
# @lc code=end

File: 2958.length-of-longest-subarray-with-at-most-k-frequency.java
/*
 * @lc app=leetcode id=2958 lang=java
 *
 * [2958] Length of Longest Subarray With at Most K Frequency
 */
import java.util.*;
// @lc code=start
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int left = 0, ans = 0, n = nums.length;
        for(int right = 0; right < n; right++){
            map.put(nums[right], map.getOrDefault(nums[right], 0) + 1);
            while(map.get(nums[right]) > k){
                int leftNum = nums[left];
                map.put(leftNum, map.get(leftNum) - 1);
                left++;
            }
            ans = Math.max(ans, right - left + 1);

        }
        return ans;
    }
}
// @lc code=end

